contracts/pool/src/contract.rs

```diff
@@ -1078,29 +1080,91 @@ pub fn assert_max_spread(env: &Env, max_spread: Decimal, return_amount: i128, sp
 /// - The commission amount, representing the fees charged for the swap.
 /// - The referral comission fee.
 pub fn compute_swap(
+    env: &Env,
     offer_pool: i128,
     ask_pool: i128,
     offer_amount: i128,
     commission_rate: Decimal,
     referral_fee: i64,
 ) -> ComputeSwap {
+    // To avoid any overflow errors we convert the input parameters to U256;
+    let offer_pool_as_u256 = U256::from_u128(env, offer_pool as u128);
+    let ask_pool_as_u256 = U256::from_u128(env, ask_pool as u128);
+    let offer_amount_as_u256 = U256::from_u128(env, offer_amount as u128);
+
     // Calculate the cross product of offer_pool and ask_pool
-    let cp: i128 = offer_pool * ask_pool;
+    let cp: U256 = offer_pool_as_u256.mul(&ask_pool_as_u256);
 
     // Calculate the resulting amount of ask assets after the swap
     // Return amount calculation based on the AMM model's invariant,
     // which ensures the product of the amounts of the two assets remains constant before and after a trade.
-    let return_amount: i128 = ask_pool - (cp / (offer_pool + offer_amount));
+
+    let return_amount: U256 =
+        ask_pool_as_u256.sub(&cp.div(&offer_pool_as_u256.add(&offer_amount_as_u256)));
+
     // Calculate the spread amount, representing the difference between the expected and actual swap amounts
-    let spread_amount: i128 = (offer_amount * ask_pool / offer_pool) - return_amount;
+    let spread_amount: U256 = (offer_amount_as_u256
+        .mul(&ask_pool_as_u256)
+        .div(&offer_pool_as_u256))
+    .sub(&return_amount);
 
-    let commission_amount: i128 = return_amount * commission_rate;
+    let commission_amount: U256 = return_amount.mul(&U256::from_u128(
+        env,
+        (commission_rate.numerator() / commission_rate.denominator()) as u128,
+    ));
 
     // Deduct the commission (minus the part that goes to the protocol) from the return amount
-    let return_amount: i128 = return_amount - commission_amount;
-    let referral_fee_amount: i128 = return_amount * Decimal::bps(referral_fee);
+    let return_amount: U256 = return_amount.sub(&commission_amount);
+
+    let decimal_bps_u256 = U256::from_u128(env, (referral_fee * 100_000_000_000_000) as u128);
+    let referral_fee_amount = return_amount.mul(&decimal_bps_u256);
+
+    let return_amount: U256 = return_amount.sub(&referral_fee_amount);
+
+    // we now convert all the results back to i128 safely
+    let return_amount = match return_amount.to_u128() {
+        Some(return_amount) => i128::try_from(return_amount).unwrap_or_else(|_| {
+            log!(env, "Pool: compute swap: cannot convert value to i128");
+            panic_with_error!(env, ContractError::CannotConvertToI128);
+        }),
+        None => {
+            log!(env, "Pool: compute swap: cannot convert value to u128");
+            panic_with_error!(env, ContractError::CannotConvertToU128);
+        }
+    };
```

Comment:
If you have to repeat this converison, this would be a good place for a helper function.
```rust
fn u256_to_i128(env: &Env, value: U256, error_msg: &str) -> i128 {
    value.to_u128()
        .and_then(|v| i128::try_from(v).ok())
        .unwrap_or_else(|| {
            log!(env, error_msg);
            panic_with_error!(env, ContractError::CannotConvertToI128);
        })
}
```

Reply:
And following:
```rust
let offer_pool_as_u256 = U256::from_u128(env, offer_pool as u128);
let ask_pool_as_u256 = U256::from_u128(env, ask_pool as u128);
let offer_amount_as_u256 = U256::from_u128(env, offer_amount as u128);

// Calculate the cross product of offer_pool and ask_pool
let cp: U256 = offer_pool_as_u256.mul(&ask_pool_as_u256);

// Calculate the return amount
let return_amount_u256 = ask_pool_as_u256.sub(&cp.div(offer_pool_as_u256.add(offer_amount_as_u256)));

// Calculate spread amount
let spread_amount_u256 = offer_amount_as_u256
    .mul(ask_pool_as_u256)
    .div(offer_pool_as_u256.add(offer_amount_as_u256))
    .sub(&return_amount_u256);

// Calculate commission amount
let commission_rate_u256 = U256::from_u128(env, (commission_rate.numerator() / commission_rate.denominator()) as u128);
let commission_amount_u256 = return_amount_u256.mul(&commission_rate_u256);

// Subtract commission amount from return amount
let return_amount_after_commission_u256 = return_amount_u256.sub(&commission_amount_u256);

// Calculate referral fee amount
let referral_fee_rate_u256 = U256::from_u128(env, (referral_fee * 100_000_000_000_000) as u128);
let referral_fee_amount_u256 = return_amount_after_commission_u256.mul(&referral_fee_rate_u256);

// Subtract referral fee amount from return amount
let final_return_amount_u256 = return_amount_after_commission_u256.sub(&referral_fee_amount_u256);

// Convert all results back to i128 safely
let return_amount_i128 = u256_to_i128(env, final_return_amount_u256, "Pool: compute swap: cannot convert return amount to i128");
let spread_amount_i128 = u256_to_i128(env, spread_amount_u256, "Pool: compute swap: cannot convert spread amount to i128");
let commission_amount_i128 = u256_to_i128(env, commission_amount_u256, "Pool: compute swap: cannot convert commission amount to i128");
let referral_fee_amount_i128 = u256_to_i128(env, referral_fee_amount_u256, "Pool: compute swap: cannot convert referral fee amount to i128");
```